dimod/include/dimod/adjvectordqm.h

```diff
@@ -0,0 +1,356 @@
+        assert(v >= 0 && v < this->num_variables());
+        assert(case_v >= 0 && case_v < this->num_cases(v));
+        if (u == v) {
+            return false;
+        }
+        variable_type cu = case_starts_[u] + case_u;
+        variable_type cv = case_starts_[v] + case_v;
+        bqm_.set_quadratic(cu, cv, bias);
+        connect_variables(u, v);
+        return true;
+    }
+
+    // Returns false if there is no interaction among the variables.
+    template <class io_bias_type>
+    bool get_quadratic(variable_type u, variable_type v,
+                       io_bias_type *quadratic_biases) {
+        assert(u >= 0 && u < this->num_variables());
+        assert(v >= 0 && v < this->num_variables());
+        if (!connection_present(u, v)) {
+            return false;
+        }
+        variable_type num_cases_u = num_cases(u);
+        variable_type num_cases_v = num_cases(v);
+        memset(
+            quadratic_biases, 0,
+            (size_t)num_cases_u * (size_t)num_cases_v * sizeof(io_bias_type));
+#pragma omp parallel for
+        for (variable_type case_u = 0; case_u < num_cases_u; case_u++) {
+            auto span =
+                bqm_.neighborhood(case_starts_[u] + case_u, case_starts_[v]);
+            while (span.first != span.second &&
+                   (span.first)->first < case_starts_[v + 1]) {
+                variable_type case_v = (span.first)->first - case_starts_[v];
+                quadratic_biases[case_u * num_cases_v + case_v] =
+                    (span.first)->second;
+                span.first++;
+            }
+        }
+        return true;
+    }
+
+    template <class io_bias_type>
+    bool set_quadratic(variable_type u, variable_type v, io_bias_type *biases) {
+        assert(u >= 0 && u < this->num_variables());
+        assert(v >= 0 && v < this->num_variables());
+        if (u == v) {
+            return false;
+        }
+        variable_type num_cases_u = num_cases(u);
+        variable_type num_cases_v = num_cases(v);
+        // This cannot be parallelized since the vectors cannot be reshaped in
+        // parallel.
+        bool inserted = false;
+        for (variable_type case_u = 0; case_u < num_cases_u; case_u++) {
+            variable_type cu = case_starts_[u] + case_u;
+            for (variable_type case_v = 0; case_v < num_cases_v; case_v++) {
+                variable_type cv = case_starts_[v] + case_v;
+                bias_type bias = biases[case_u * num_cases_v + case_v];
+                if (bias != (bias_type)0) {
+                    bqm_.set_quadratic(cu, cv, bias);
+                    inserted = true;
+                }
+            }
+        }
+
+        if (inserted) {
+            connect_variables(u, v);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    template <class io_variable_type, class io_bias_type>
+    void get_energies(io_variable_type *samples, int num_samples,
```

Comment:
IMO it makes sense to just have a single method, `bias_type energy(Iter sample_start)`. The user can then parallelize it themselves if they like.